client-apis/decrypt.md

```diff
@@ -69,65 +98,259 @@ This default CMM MUST obtain the decryption materials required for decryption.
+from the [keyring](../framework/keyring-interface.md) inputted.
+
+The call to the CMM's [Decrypt Materials](../framework/cmm-interface.md#decrypt-materials) operation
+MUST be constructed as follows:
+
+- Encryption Context: This is the parsed [encryption context](../data-format/message-header.md#aad)
+  from the message header.
+- Algorithm Suite ID: This is the parsed
+  [algorithm suite ID](../data-format/message-header.md#algorithm-suite-id)
+  from the message header.
+- Encrypted Data Keys: This is the parsed [encrypted data keys](../data-format/message-header#encrypted-data-keys)
+  from the message hedaer.
+
+The data key used as input for all decryption described below is a data key derived from the plaintext data key
+included in the [decryption materials](../framework/structures.md#decryption-materials).
+The algorithm used to derive a data key from the plaintext data key MUST be
+the [key derivation algorithm](../framework/algorithm-suites.md#key-derivation-algorithm) included in the
+[algorithm suite](../framework/algorithm-suites.md) defined above.
+This document refers to the output of the key derivation algorithm as the derived data key.
+Note that if the key derivation algorithm is the [identity KDF](../framework/algorithm-suites.md#identity-kdf),
+then the derived data key is the same as the plaintext data key.
+
+### Verify the header
+
+Once a valid message header is deserialized and decryption materials are available,
+this operation MUST validate the [message header body](../data-format/message-header.md#header-body)
+by using the [authenticated encryption algorithm](../framework/algorithm-suites.md#encryption-algorithm)
+to decrypt with the following inputs:
+
+- the AAD is the serialized [message header body](../data-format/message-header.md#header-body).
+- the IV is the value serialized in the message header's [IV field](../data-format/message-header#iv).
+- the cipherkey is the derived data key
+- the ciphertext is an empty byte array
+
+If this tag verification fails, this operation MUST immediately halt and fail.
+
+If the input encrypted message is being streamed to this operation:
+
+- This operation SHOULD release the parsed [encryption context](#encryption-context),
+  [algorithm suite ID](#algorithm-suite-id),
+  and [other header information](#parsed-header)
+  as soon as tag verification succeeds.
+  However, if this operation is using an algorithm suite with a signature algorithm
+  all released output MUST be not considered verified until
+  this operation successfully completes.
+  See [security considerations](#security-considerations) below.
+- This operation SHOULD input the serialized header to the signature algorithm as soon as it is deserialized,
+  such that the serialized frame isn't required to remain in memory to complete
+  the [signature calculation](#signature-calculation).
+
+### Decrypt the message body
+
```

Comment:
Implied by some other bits of this, but the plaintext of a frame MUST NOT be released until it has been authenticated.